atinout.1.ronn
atinout(1) -- Send AT commands to modem, capturing the response
===============================================================

## SYNOPSIS

`atinout` <input_file>|`-` <modem_device> <output_file>|`-`<br>
`atinout` `--version`<br>
`atinout` `--usage`<br>
`atinout` `--help`<br>

## DESCRIPTION

**Atinout** reads a list of AT commands. It sends those commands one by one
to the modem, waiting for the final result code for the
currently running command before continuing with the next command in
the list. The output from the commands is saved.

Notice that **atinout** is strictly about sending AT commands only; it
cannot be used for AT commands that expect additional input data (e.g.
AT+CMGS etc).

## FILES

**Atinout** expects <input_file> to be a list of AT command to be run. If
the file name `-` is used then standard input is read.

Although the above states a list of AT commands it is actually not just
a list of command but more generally a list of AT command lines. However
the rest of the man page will just refer to AT commands; running several
commands in one command line offers no benefits, and if one of them fails
you have no idea of which one failed. So unless maybe for a single run
on the command line (see EXAMPLES), I recommend against putting more
than one command on each command line.

And if the distinction between AT command and AT command line
was not something you were deeply familiar with you should read
the ITU V.250 specification about AT commands. In fact, unless
you already have read V.250 before, you absolutely should go to
http://www.itu.int/rec/T-REC-V.250-200307-I/en and do so right away.

The modem is identified with <modem_device> (e.g. /dev/ttyS0 (COM1),
/dev/ttyACM0 (USB phone serial interface), ...). No serial configuration
(e.g. speed, parity) is done before opening the device.

The responses from the AT commands are written to <output_file>, or
standard output if the file name `-` is given.

## OPTIONS

The following list of options are supported:

  * `-h`, `--help`:
    Prints explanation and gives an example on usage.

  * `--usage`:
    Prints information about arguments and options.

  * `--version`:
    Print version information. If this is the only option given, also prints
    copyright/licensing information.


## EXAMPLES

Check that the modem is connected and alive.

    $ echo AT | atinout - /dev/ttyACM0 -
    AT
    OK
    $

Hang up any calls.

    $ echo ATH | atinout - /dev/ttyACM0 -
    ATH
    OK
    $

Query how many entries the phone book has for ME storage.

    $ echo 'AT+CPBS="ME"; +CPBR=?' | atinout - /dev/ttyACM0 -
    AT+CPBS="ME"; +CPBR=?
    
    +CPBR: (1-7000),80,62
    
    OK
    $

Backup all phone numbers. This will most likely not be the same as
a full backup of all contact information, but at least the names and
numbers are saved which is the most vital information.

    $ cat > save-phonebooks <<EOF
    AT+CSCS="UTF-8"
    AT+CGMI
    AT+CGMR
    AT+CPBS="SM"
    AT+CPBR=1,240
    AT+CPBS="ME"
    AT+CPBR=1,7000
    EOF
    $ atinout save-phonebooks /dev/ttyACM0 phonebook_backup.`date "+%Y-%m-%d_%H-%M"`
    $

List current calls. The information text format is (see 27.007 for details):

    +CLCC: <id1>,<dir>,<stat>,<mode>,<mpty>[,<number>,<type>[,<alpha>[,<priority>]]]

    # No active calls
    $ echo AT+CLCC | atinout - /dev/ttyACM0 -
    AT+CLCC
    OK
    $

    # One active call
    $ echo AT+CLCC | atinout - /dev/ttyACM0 -
    AT+CLCC
    +CLCC: 1,0,3,0,0,"0123456789",129,"",0
    
    OK
    $

## ENVIRONMENT

No environmental variables are used.


## BUGS

**Atinout** does not support reading response lines longer than 4095
characters.

**Atinout** does not do any serial port configuration, it just opens
the device as is.

## COPYRIGHT

Atinout is Copyright (C) 2013 Håkon Løvdal <hlovdal@users.sourceforge.net>

## SEE ALSO

V.250 (http://www.itu.int/rec/T-REC-V.250-200307-I/en),
27.007 (http://www.3gpp.org/ftp/Specs/html-info/27007.htm),
27.005 (http://www.3gpp.org/ftp/Specs/html-info/27005.htm)



Changelog

Atinout changelog
=================

Version 0.9.1 (2013-09-08)
----------------------------

* Add RPM spec file, "rpmbuild -ta atinout-0.9.1.tar.gz" will work.
* Fix Makefile so that it works without problems when compiling source
from tar file (git source was always ok).
* Fixed test checking wrong file when testing if opnening it was ok.


Version 0.9 (2013-04-24)
----------------------------

* Initial release.



atinout/atinout.c
/*
 * atinout
 *
 * This program takes AT commands as input, sends them to the modem and outputs
 * the responses.
 *
 * Copyright (C) 2013 Håkon Løvdal <hlovdal@users.sourceforge.net>
 *
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <errno.h>
#include <getopt.h>

#define MAX_LINE_LENGTH (4 * 1024)
static char buf[MAX_LINE_LENGTH];

static struct option long_options[] = {
    {"help", no_argument, NULL, 'h'},
    {"version", no_argument, NULL, 'V'},
    {"usage", no_argument, NULL, 0},
    {NULL, 0, NULL, 0}
};
static const char *short_options = "hV";

static void usage(const char * const argv0)
{
    printf("Usage: %s <input_file> <modem_device> <output_file>\n", argv0);
    printf("\n");
    printf("\t<input_file> is a file with AT commands to be executed. Value '-' means standard input.\n");
    printf("\t<modem_device> is the device file for the modem (e.g. /dev/ttyACM0, /dev/ttyS0, etc).\n");
    printf("\t<output_file> is a file the responses to the AT commands are saved. Value '-' means standard output.\n");
    printf("\n");
    printf("In addition the program supports the following options as the only argument:\n");
    printf("\n");
    printf("\t-h|--help\n");
    printf("\t-V|--version\n");
    printf("\t--usage\n");
    printf("\n");
}

static void help(const char * const argv0)
{
    printf("This program takes AT commands as input, sends them to the modem and outputs\n");
    printf("the responses.\n");
    printf("\n");
    printf("Example:\n");
    printf("\n");
    printf("$ echo 'AT+CPBS=\"ME\",\"ME\"; +CPBR=?' | %s - /dev/ttyACM0 -\n", argv0);
    printf("AT+CPBS=\"ME\",\"ME\"; +CPBR=?\n");
    printf("\n");
    printf("+CPBR: (1-7000),80,62\n");
    printf("\n");
    printf("OK\n");
    printf("$\n");
    printf("\n");
    printf("See also the --usage command line option\n");

}

/**
 * Replace '\n' with '\r', aka `tr '\012' '\015'` for a single line.
 * Supports both "\n" and "\r\n" line endings.
 */
static bool tr_lf_cr(const char *s)
{
    char *p;
    p = strchr(s, '\n');
    if (p == NULL || p[1] != '\0') {
        return false;
    }
    if (p > s && p[-1] == '\r') {
        *p = '\0';
    } else {
        *p = '\r';
    }
    return true;
}

static void strip_cr(char *s)
{
    char *from, *to;
    from = to = s;
    while (*from != '\0') {
        if (*from == '\r') {
            from++;
            continue;
        }
        *to++ = *from++;
    }
    *to = '\0';
}

static bool is_final_result(const char * const response)
{
#define STARTS_WITH(a, b) ( strncmp((a), (b), strlen(b)) == 0)
    switch (response[0]) {
        case '+':
            if (STARTS_WITH(&response[1], "CME ERROR:")) {
                return true;
            }
            if (STARTS_WITH(&response[1], "CMS ERROR:")) {
                return true;
            }
            return false;
        case 'B':
            if (strcmp(&response[1], "USY\n") == 0) {
                return true;
            }

            return false;

        case 'E':
            if (strcmp(&response[1], "RROR\n") == 0) {
                return true;
            }
            return false;
        case 'N':
            if (strcmp(&response[1], "O ANSWER\n") == 0) {
                return true;
            }
            if (strcmp(&response[1], "O CARRIER\n") == 0) {
                return true;
            }
            if (strcmp(&response[1], "O DIALTONE\n") == 0) {
                return true;
            }
            return false;
        case 'O':
            if (strcmp(&response[1], "K\n") == 0) {
                return true;
            }
            // fallthrough
        default:
            return false;
    }
}

int main(int argc, char *argv[])
{
    FILE *atcmds;
    FILE *modem;
    FILE *output;
    char *line;
    bool success;
    int res;
    int option_index = 0;
    int c;

    while ((c = getopt_long(argc, argv, short_options, long_options, &option_index)) != -1) {
        switch (c) {
            case 'h':
                help(argv[0]);
                return EXIT_SUCCESS;
            case 'V':
                printf("atinout version " VERSION "\n");
                if (argc == 2) {
                    printf("Copyright (C) 2013 Håkon Løvdal <hlovdal@users.sourceforge.net>\n"
                            "This program comes with ABSOLUTELY NO WARRANTY.\n"
                            "This is free software, and you are welcome to redistribute it under\n"
                            "certain conditions; see http://www.gnu.org/licenses/gpl.html for details.\n");
                    return EXIT_SUCCESS;
                }
                break;
            case 0:
                if (strcmp("usage", long_options[option_index].name) == 0) {
                    usage(argv[0]);
                    return EXIT_SUCCESS;
                }
                break;
            case '?':
                break;
            default:
                fprintf(stderr, "getopt returned character code 0x%02x\n", (unsigned int)c);
        }
    }

    if ((argc - optind) != 3) {
        usage(argv[0]);
        return EXIT_FAILURE;
    }

#define INPUT_FILE   argv[optind + 0]
#define MODEM_DEVICE argv[optind + 1]
#define OUTPUT_FILE  argv[optind + 2]

    if (strcmp(INPUT_FILE, "-") == 0) {
        atcmds = stdin;
    } else {
        atcmds = fopen(INPUT_FILE, "rb");
        if (atcmds == NULL) {
            fprintf(stderr, "fopen(%s) failed: %s\n", INPUT_FILE, strerror(errno));
            return EXIT_FAILURE;
        }
    }

    modem = fopen(MODEM_DEVICE, "r+b");
    if (modem == NULL) {
        fprintf(stderr, "fopen(%s) failed: %s\n", MODEM_DEVICE, strerror(errno));
        return EXIT_FAILURE;
    }

    if (strcmp(OUTPUT_FILE, "-") == 0) {
        output = stdout;
    } else {
        output = fopen(OUTPUT_FILE, "wb");
        if (output == NULL) {
            fprintf(stderr, "fopen(%s) failed: %s\n", OUTPUT_FILE, strerror(errno));
            return EXIT_FAILURE;
        }
    }

    while ((line = fgets(buf, (int)sizeof(buf), atcmds)) != NULL) {
        success = tr_lf_cr(line);
        if (! success) {
            fprintf(stderr, "invalid string: '%s'\n", line);
            return EXIT_FAILURE;
        }
        res = fputs(line, modem);
        if (res < 0) {
            fprintf(stderr, "failed to send '%s' to modem (res = %d)\n", line, res);
            return EXIT_FAILURE;
        }
        do {
            line = fgets(buf, (int)sizeof(buf), modem);
            if (line == NULL) {
                fprintf(stderr, "EOF from modem\n");
                return EXIT_FAILURE;
            }
            strip_cr(line);
            res = fputs(line, output);
            if (res < 0) {
                fprintf(stderr, "failed to write '%s' to output file (res = %d)\n", line, res);
                return EXIT_FAILURE;
            }
        } while (! is_final_result(line));
    }

    if (strcmp(OUTPUT_FILE, "-") != 0) {
        res = fclose(output);
        if (res != 0) {
            fprintf(stderr, "closing output failed: %s\n", strerror(errno));
            return EXIT_FAILURE;
        }
    }
    res = fclose(modem);
    if (res != 0) {
        fprintf(stderr, "closing modem failed: %s\n", strerror(errno));
        return EXIT_FAILURE;
    }
    if (strcmp(INPUT_FILE, "-") != 0) {
        res = fclose(atcmds);
        if (res != 0) {
            fprintf(stderr, "closing input failed: %s\n", strerror(errno));
            return EXIT_FAILURE;
        }
    }
    return EXIT_SUCCESS;
}

